agent/websockets/connection.go

```diff
@@ -252,3 +264,47 @@ func (conn *Connection) ReadServerMessages() ([]interface{}, error) {
 		return nil, nil
 	}
 }
+
+// injectWebsocketMessage injects a shim header value into a single websocket message in-place.
+// Returns a pointer to a new copy of the struct on success.
+func injectWebsocketMessage(msg *message, injectionPath []string, injectionValues map[string]string) (*message, error) {
+	if msg == nil || injectionValues == nil {
+		return nil, errors.New("unexpected nil message")
+	}
+	if injectionValues == nil {
+		return nil, errors.New("unexpected nil values map")
+	}
+	if len(injectionValues) == 0 {
+		return msg, nil
+	}
+	// Deserialize the websocket message into a JSON object.
+	var origJSONComponent map[string]interface{}
+	err := json.Unmarshal(msg.Data, &origJSONComponent)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal as json message: %v", err)
+	}
+	var currJSONComponent map[string]interface{}
+	var ok bool
+	currJSONComponent = origJSONComponent
+	for _, pathComponent := range injectionPath {
+		currJSONComponent, ok = currJSONComponent[pathComponent].(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("failed to inject value due to unexpected message json format")
+		}
+	}
+
+	// Inject the header values into the specified key. Do not overwrite existing header values.
```

Comment:
Is it OK to allow passing values both in headers and message body? This way client can put anything into any header, potentially overriding e.g. reverse proxy values like X-Real-IP if any.

Reply:
In our case yes, because we are injecting the headers into a websocket message rather than the other way around. Additionally, existing headers are explicitly checked and not overwritten, and finally, the server receiving the websocket message is the one which should only check message payload headers which are guaranteed to be set by the client.

In CDAP's case, the only two headers which could be set are the Authorization and Content-Type headers which should be low risk.